src/analysis/normed_space/basic.lean

```diff
@@ -720,18 +720,39 @@ lemma tendsto_norm_nhds_within_zero : tendsto (norm : α → ℝ) (𝓝[{0}ᶜ]
 
 end normed_group
 
-section normed_ring
+section semi_normed_ring
+
+/-- A seminormed ring is a ring endowed with a seminorm which satisfies the inequality
+`∥x y∥ ≤ ∥x∥ ∥y∥`. -/
+class semi_normed_ring (α : Type*) extends has_norm α, ring α, pseudo_metric_space α :=
+(dist_eq : ∀ x y, dist x y = norm (x - y))
+(norm_mul : ∀ a b, norm (a * b) ≤ norm a * norm b)
 
 /-- A normed ring is a ring endowed with a norm which satisfies the inequality `∥x y∥ ≤ ∥x∥ ∥y∥`. -/
 class normed_ring (α : Type*) extends has_norm α, ring α, metric_space α :=
 (dist_eq : ∀ x y, dist x y = norm (x - y))
 (norm_mul : ∀ a b, norm (a * b) ≤ norm a * norm b)
```

Comment:
Why aren't you using `extends semi_normed_ring α, metric_space α`?

Reply:
I followed the same approach used for `semi_normed_group`, to copy the definition of `normed_group` and add the instance `normed_to_semi_normed` manually. It has been suggested on Zulip [here](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/seminormed.20group). If I use
```
set_option old_structure_cmd true
/-- A normed ring is a ring endowed with a norm which satisfies the inequality `∥x y∥ ≤ ∥x∥ ∥y∥`. -/
class normed_ring (α : Type*) extends semi_normed_ring α, ring α, metric_space α
```
Then I have troubles to prove `instance normed_ring.to_normed_group [β : normed_ring α] : normed_group α`: the following
```
@[priority 100] -- see Note [lower instance priority]
instance normed_ring.to_normed_group [β : normed_ring α] : normed_group α := { 
  dist_eq := λ x y,
  begin
    exact semi_normed_ring.dist_eq x y,
  end }
```
gives
```
invalid type ascription, term has type
  @dist α
      (@pseudo_metric_space.to_has_dist α
         (@semi_normed_ring.to_pseudo_metric_space α (@normed_ring.to_semi_normed_ring α β)))
      x y = ∥x - y∥
but is expected to have type
  @dist α
      (@pseudo_metric_space.to_has_dist α
         (@metric_space.to_pseudo_metric_space α (@normed_ring.to_metric_space α β)))
      x  y = ∥x - y∥
```

Reply:
OK, I tried
```
set_option old_structure_cmd true
/-- A normed ring is a ring endowed with a norm which satisfies the inequality `∥x y∥ ≤ ∥x∥ ∥y∥`. -/
class normed_ring (α : Type*) extends semi_normed_ring α, ring α, metric_space α
```
and
```
/-- A normed commutative ring is a commutative ring endowed with a norm which satisfies
the inequality `∥x y∥ ≤ ∥x∥ ∥y∥`. -/
class normed_comm_ring (α : Type*) extends semi_normed_comm_ring α, metric_space α
```
I don't know exactly what `set_option old_structure_cmd true` does, but now the constructor is different, right? For example, the proof of `instance : normed_comm_ring ℤ` should be changed. I can do it if you think this other approach is better.

Reply:
It just feels ever so slightly duplicative to have to write the axiom out of second time. But it's not a big deal, and we use the pattern extensively elsewhere, so I'm happy to leave it as is.